Cloud Platform (#64)
OctopusDeploy/Architecture: CloudPlatform/ExternalServices.md

# Summary # 

> Cloud Platform's goal to provide a rock solid, fully managed hosting platform for Octopus Deploy customers. It achieves this by orchestrating a group of reliable external services and filling any remaining gaps with custom code.

Reviewer: goal to -> goal is to

Author: 👍

> Cloud Platform's goal to provide a rock solid, fully managed hosting platform for Octopus Deploy customers. It achieves this by orchestrating a group of reliable external services and filling any remaining gaps with custom code.

Reviewer: ```suggestion
Cloud Platform's goal is to provide a rock-solid, fully managed hosting platform for Octopus Deploy customers. It achieves this by orchestrating a group of reliable external services and filling any remaining gaps with custom code.
```

In most cases the custom code is meant to be relatively short lived and removed once one of the external services provides a feature that replaces it.

# Details #

> At this stage of Cloud Platform (CP) development (e.g. we are not Netflix yet) we avoid reinviting the wheel where that doesn't directly benefit CP.  E.g. We could hire bunch of DBAs and host our own SQL Servers in Azure but decided against that because Azure SQL can provide that service for us reliably. 

Reviewer 2: ```suggestion
At this stage of Cloud Platform (CP) development (e.g. we are not Netflix yet) we avoid reinventing the wheel where that doesn't directly benefit CP.  E.g. We could hire bunch of DBAs and host our own SQL Servers in Azure but decided against that because Azure SQL can provide that service for us reliably. 
```

The same applies to Azure Storage, and Azure Kubernetes Service (AKS), etc.

# Filling The Gaps #

When we started using AKS it didn't provide [autoscaler](https://github.com/kubernetes/autoscaler) out of the box. We had to deploy our own which worked but a) took a bit of our time b) we hit bunch of issues as the version we were using was not working 100% correctly with the version of K8s we were on. After a year or so AKS included autoscaler as an option and now AKS team makes sure that they ship the correct version of autoscaler with each version of K8s they offer.  This allowed us to delete our code. 

When we started neither K8s nor any of the Azure services executing containers provided hard walls between containers which is what we needed to run arbitrary code from our customers. As a result we developed a custom solution called Dynamic Workers. 

> 3 years later, [Azure Container Instances] seems to offer this feature which is why we consider moving there.

Reviewer: ```suggestion
3 years later, [Azure Container Instances] seems to offer this feature which is why we're considering moving there.
```
